File: Practice/Graphs/PathInDirectedGraph.js
function PathInDirectedGraph(A, B) {
  let sourceNode = 1;
  let hs = new Set();
  let queue = [];
  queue.push(sourceNode);
  hs.add(sourceNode);

  let hm = {};

  for (let i = 1; i <= A; i++) {
    hm[i] = [];
  }

  for (let i = 0; i < B.length; i++) {
    let valArray = hm[B[i][0]];
    valArray.push(B[i][1]);
    hm[B[i][0]] = valArray;
  }

  while (queue.length > 0) {
    let currNode = queue.shift();
    for (let i = 0; i < hm[currNode].length; i++) {
      if (hs.has(hm[currNode][i])) continue;
      let val = hm[currNode][i];
      queue.push(val);
      hs.add(val);
    }
  }

  if (hs.has(A)) return 1;
  return 0;
}

A = 5;
B = [
  [1, 2],
  [4, 1],
  [2, 4],
  [3, 4],
  [5, 2],
  [1, 3],
];

console.log(PathInDirectedGraph(A, B));

File: Revision/Graphs/PathInDirectedGraph.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Queue;

public class PathInDirectedGraph {
    public int solve(int A, ArrayList<ArrayList<Integer>> B) {
        // Creating adjacency list
        HashMap<Integer, ArrayList<Integer>> adjacencyMap = new HashMap<>();
        createAdjacencyList(B, adjacencyMap);
        // check if 1 has a path to A through breath first search
        return bfs(1, adjacencyMap, A);
    }

    public int bfs(int node, HashMap<Integer, ArrayList<Integer>> adjacencyMap, int A) {
        Queue<Integer> bfsQueue = new LinkedList<>();
        bfsQueue.offer(node);
        while (!bfsQueue.isEmpty()) {
            int currentValue = bfsQueue.poll();
            if (adjacencyMap.containsKey(currentValue)) {
                ArrayList<Integer> valueList = adjacencyMap.get(currentValue);
                for (int i = 0; i < valueList.size(); i++) {
                    int value = valueList.get(i);
                    if (value == A)
                        return 1;
                    bfsQueue.add(value);
                }
            }
        }

        return 0;
    }

    public void createAdjacencyList(ArrayList<ArrayList<Integer>> B,
            HashMap<Integer, ArrayList<Integer>> adjacencyMap) {
        for (int i = 0; i < B.size(); i++) {
            int currentKey = B.get(i).get(0);
            int currentValue = B.get(i).get(1);
            ArrayList<Integer> currentValueList;
            if (adjacencyMap.containsKey(currentKey)) {
                currentValueList = adjacencyMap.get(currentKey);
                currentValueList.add(currentValue);

            } else {
                currentValueList = new ArrayList<>();
                currentValueList.add(currentValue);
            }
            adjacencyMap.put(currentKey, currentValueList);

        }
    }
}